File: Problem_Sets_500_599/DC_561_etsy_hard/array_to_binary_tree.js
// This problem was asked by Etsy.

// Given a sorted array, convert it into a height-balanced binary search tree.

const { Node } = require("../../Data_Structures/Binary_Tree/binary_tree.js")

function array_to_binary(array) {
    // empty array return null    
    if (array.length == 0) {
        return null
    }
    if (array.length == 1) {
        let node = new Node(array[0])
        return node
    }
    else {
        let center_index = Math.floor(array.length / 2)
        let left_tree = array.slice(0, center_index)
        let right_tree = array.slice(center_index + 1)
        let node = new Node(array[center_index])
        if (left_tree.length > 0) {
            node.left = array_to_binary(left_tree)
        }
        if (right_tree.length > 0) {
            node.right = array_to_binary(right_tree)
        }
        return node
    }
}


//↓↓↓   basic test   ↓↓↓
function testing() {
    let sorted1 = [2, 3, 4, 5, 6, 7, 8]
    let sorted2 = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
    let sorted3 = [1]
    let sorted4 = []

    let binary_tree1 = array_to_binary(sorted1)
    let binary_tree2 = array_to_binary(sorted2)

    console.log(binary_tree1)
    console.log(binary_tree2)

    let binary_tree3 = array_to_binary(sorted3)
    let binary_tree4 = array_to_binary(sorted4)

    console.log(binary_tree3)
    console.log(binary_tree4)
}

testing();

File: Problem_Sets_500_599/DC_561_etsy_hard/array_to_binary_tree.py
# This problem was asked by Etsy.

# Given a sorted array / list, convert it into a height-balanced binary search tree.

from binary_tree import Node
import math

def list_to_binary(list):
    if len(list) == 0 :
        return None
    if len(list) == 1 :
        node = Node(list[0])
        return node
    else :
        center_index = math.floor(len(list) / 2)
        left_tree = list[ : center_index ]
        right_tree = list[ center_index + 1 : ]
        node = Node(list[center_index])
        if len(left_tree) > 0 :
            node.left = list_to_binary(left_tree)
        if len(right_tree) > 0 :
            node.right = list_to_binary(right_tree)
        return node
    



# ↓↓↓   basic test   ↓↓↓

sorted0 = [1,2,3]

binary_tree0 = list_to_binary(sorted0)
print(binary_tree0.data, binary_tree0.left.data, binary_tree0.right.data)
# 2 1 3

sorted1 = [2,3,4,5,6,7,8] 

sorted2 = [0,1,2,3,4,5,6,7,8,9] 

binary_tree1 = list_to_binary(sorted1)
binary_tree2 = list_to_binary(sorted2)

print(binary_tree1.data, binary_tree1.left.data, binary_tree1.right.data)
# 5 3 7


print(binary_tree2.data, binary_tree2.left.data, binary_tree2.right.data)
# 5 2 8